runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainer.java

```diff
@@ -0,0 +1,288 @@
+      new Function<MetricAggregator<?>, DistributionAggregator>() {
+        @Override
+        public DistributionAggregator apply(MetricAggregator<?> metricAggregator) {
+          return (DistributionAggregator) metricAggregator;
+        }
+      };
+
+  SparkMetricsContainer merge(SparkMetricsContainer other) {
+    return
+        new SparkMetricsContainer()
+            .updated(this.getAggregators())
+            .updated(other.getAggregators());
+  }
+
+  private Collection<MetricAggregator<?>> getAggregators() {
+    return metrics.values();
+  }
+
+  private void writeObject(ObjectOutputStream out) throws IOException {
+    materialize();
+    out.defaultWriteObject();
+  }
+
+  private void materialize() {
+    if (metricsContainers != null) {
+      for (MetricsContainer container : metricsContainers.asMap().values()) {
+        MetricUpdates cumulative = container.getCumulative();
+        updated(Iterables.transform(cumulative.counterUpdates(), TO_COUNTER_AGGREGATOR));
+        updated(Iterables.transform(cumulative.distributionUpdates(), TO_DISTRIBUTION_AGGREGATOR));
+      }
+    }
+  }
+
+  private static final Function<MetricUpdate<Long>, MetricAggregator<?>>
+      TO_COUNTER_AGGREGATOR = new Function<MetricUpdate<Long>, MetricAggregator<?>>() {
+    @SuppressWarnings("ConstantConditions")
+    @Override
+    public CounterAggregator
+    apply(MetricUpdate<Long> update) {
+      return update != null ? new CounterAggregator(new SparkMetricKey(update.getKey()),
+          update.getUpdate()) : null;
+    }
+  };
+
+  private static final Function<MetricUpdate<DistributionData>, MetricAggregator<?>>
+      TO_DISTRIBUTION_AGGREGATOR =
+      new Function<MetricUpdate<DistributionData>, MetricAggregator<?>>() {
+        @SuppressWarnings("ConstantConditions")
+        @Override
+        public DistributionAggregator
+        apply(MetricUpdate<DistributionData> update) {
+          return update != null ? new DistributionAggregator(new SparkMetricKey(update.getKey()),
+              new SparkDistributionData(update.getUpdate())) : null;
+        }
+      };
+
+  private SparkMetricsContainer updated(Iterable<MetricAggregator<?>> updates) {
+    for (MetricAggregator<?> update : updates) {
+      MetricKey key = update.getKey();
+      MetricAggregator<?> current = metrics.get(key);
+      Object updateValue = update.getValue();
+      metrics.put(new SparkMetricKey(key),
+          current != null ? MetricAggregator.updated(current, updateValue) : update);
+    }
+    return this;
+  }
+
+  private static class MetricsContainerCacheLoader extends CacheLoader<String, MetricsContainer> {
+    @SuppressWarnings("NullableProblems")
+    @Override
+    public MetricsContainer load(String stepName) throws Exception {
+      return new MetricsContainer(stepName);
+    }
+  }
+
+  private static class SparkMetricKey extends MetricKey implements Serializable {
```

Comment:
Same answer for these "data-only classes". The ones I'd worry about sharing and/or making serializable are the ones with logic that aren't simple autovalues meant to look like a JSON request/response.